### PR TITLE
Feature/db access rewrite

### PR DESCRIPTION
**Pull request by RainbowDashLabs**
Complete rewrite of database access

Replace the old integrated sql updater with the sql updater of sql-util
Create endpoint like layout with nested classes.
Add chained calls to make the inputs simpler.
Change method names to be more meaningful.
Add caching for high frequent calls like the current language or specific settings.
Remove prefix command and fallback listener

### CODE REVIEW

**Comment by RainbowDashLabs** [comment]
@Taucher2003 sorry in advance for this really large change, but splitting this would have been even more painful and probably not even possible.

I think we will benefit from it in the long run

File: src/main/java/de/chojo/repbot/dao/access/guild/settings/sub/Thanking.java [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+    private Thankwords thankwords;
+
+    public Thanking(Settings settings) {
+        this(settings, null, true);
+    }
+
+    public Thanking(Settings settings, String mainReaction, boolean channelWhitelist) {
+        super(settings);
+        this.settings = settings;
+        this.mainReaction = mainReaction == null ? "🏅" : mainReaction;
```
Comment (Taucher2003) [inline src/main/java/de/chojo/repbot/dao/access/guild/settings/sub/Thanking.java:37]: I think it is better to work without null here. Or is there any reason to pass null here which needs to get handled?
```suggestion
        this(settings, "🏅", true);
    }

    public Thanking(Settings settings, String mainReaction, boolean channelWhitelist) {
        super(settings);
        this.settings = settings;
        this.mainReaction = mainReaction;
```

File: src/main/java/de/chojo/repbot/dao/access/guild/settings/sub/thanking/DonorRoles.java [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+    @Override
+    public Guild guild() {
+        return thanking.guild();
+    }
+
+    @Override
+    protected String targetTable() {
+        return "donor_roles";
+    }
+
```
Comment (Taucher2003) [inline src/main/java/de/chojo/repbot/dao/access/guild/settings/sub/thanking/DonorRoles.java:25]:
```suggestion
```

File: src/main/java/de/chojo/repbot/dao/pagination/GuildRanking.java [hunk 3]
```diff
@@ -1,11 +1,15 @@
...
 
-public class GuildRanking extends PageAccess<ReputationUser> {
-    public GuildRanking(Supplier<Integer> pagecount, Function<Integer, List<ReputationUser>> pageSupplier) {
+public class GuildRanking extends PageAccess<RepProfile> {
+    public GuildRanking(Supplier<Integer> pagecount, Function<Integer, List<RepProfile>> pageSupplier) {
         super(pagecount, pageSupplier);
     }
+
+
 }
```
Comment (Taucher2003) [inline src/main/java/de/chojo/repbot/dao/pagination/GuildRanking.java:15]:
```suggestion
    }
}
```

File: src/main/java/de/chojo/repbot/listener/StateListener.java [hunk 4]
```diff
@@ -62,46 +50,43 @@ public void onGuildJoin(@NotNull GuildJoinEvent event) {
...
-        guildData.migrated(event.getGuild());
+        guilds.guild(event.getGuild()).migration().migrated();
     }
 
     @Override
     public void onGuildLeave(@NotNull GuildLeaveEvent event) {
-        if (!configuration.migration().isActive()) {
-            gdprData.queueGuildDeletion(event.getGuild());
-        }
+        if (!configuration.migration().isActive()) return;
```
Comment (Taucher2003) [inline src/main/java/de/chojo/repbot/listener/StateListener.java:58]: Fix condition
```suggestion
        if (configuration.migration().isActive()) return;
```
or like the screenshot
```suggestion
```

File: src/main/java/de/chojo/repbot/service/GdprService.java [hunk 5]
```diff
@@ -1,139 +1,75 @@
...
+            var savedIds = guilds.guild(guild).userIds();
             var memberIds = guild.loadMembers().get()
                     .stream()
-                    .map(ISnowflake::getIdLong)
-                    .collect(Collectors.toList());
-            var collect = savedIds.stream().filter(id -> !memberIds.contains(id)).collect(Collectors.toList());
-            for (var id : collect) gdprData.executeRemovalTask(new RemovalTask(-1, guild.getIdLong(), id));
+                    .map(ISnowflake::getIdLong).toList();
+            var collect = savedIds.stream().filter(id -> !memberIds.contains(id)).toList();
+            for (var id : collect) new RemovalTask(gdpr, -1, guild.getIdLong(), id);
```
Comment (Taucher2003) [inline src/main/java/de/chojo/repbot/service/GdprService.java:65]: The cleanup task also needs to be executed :D
```suggestion
            for (var id : collect) new RemovalTask(gdpr, -1, guild.getIdLong(), id).executeRemovalTask();
```

File: src/main/java/de/chojo/repbot/service/RepBotCachePolicy.java [hunk 6]
```diff
@@ -42,6 +42,10 @@ public boolean cacheMember(@NotNull Member member) {
             return true;
         }
 
+        if (member.getVoiceState().inAudioChannel()) {
+            return true;
+        }
+
```
Comment (Taucher2003) [inline src/main/java/de/chojo/repbot/service/RepBotCachePolicy.java:48]: This is already been taken care of in `MemberCachePolicy.VOICE` above
```suggestion
```

File: src/main/resources/database/postgres/1/patch_17.sql [hunk 7]
```diff
@@ -0,0 +1,5 @@
+alter table repbot_schema.guild_settings
+    drop column IF EXISTS manager_role;
+
+alter table repbot_schema.guild_settings
+    drop column IF EXISTS prefix;
```
Comment (Taucher2003) [inline src/main/resources/database/postgres/1/patch_17.sql:5]:
```suggestion
ALTER TABLE repbot_schema.guild_settings
    DROP COLUMN IF EXISTS manager_role;

ALTER TABLE repbot_schema.guild_settings
    DROP COLUMN IF EXISTS prefix;
```

File: src/main/java/de/chojo/repbot/ReputationBot.java [hunk 8]
```diff
@@ -161,24 +162,35 @@ private void initBotList() {
     }
 
     private void initDatabase() throws SQLException, IOException {
-        var connectionPool = getConnectionPool(false);
+        dataSource = getConnectionPool(true);
 
+        var updatePool = getConnectionPool(false);
         var schema = configuration.database().schema();
-        SqlUpdater.builder(connectionPool)
+        de.chojo.sqlutil.updater.SqlUpdater.builder(updatePool, SqlType.POSTGRES)
```
[the remark this answers is not in the document]
Comment (RainbowDashLabs) [reply src/main/java/de/chojo/repbot/ReputationBot.java:169]: No. Just an artifact from the migration

File: src/main/java/de/chojo/repbot/dao/access/guild/Cleanup.java [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+        builder(Boolean.class).query("""
+                                DELETE FROM self_cleanup WHERE guild_id = ?
+                        """)
+                .paramsBuilder(stmt -> stmt.setLong(guildId()))
+                .update().executeSync();
+    }
+
+    @Override
+    public Guild guild() {
+        return null;
```
[the remark this answers is not in the document]
Comment (RainbowDashLabs) [reply src/main/java/de/chojo/repbot/dao/access/guild/Cleanup.java:43]: Yeah this breaks indeed.

File: src/main/java/de/chojo/repbot/dao/access/guild/reputation/sub/RepUser.java [hunk 10]
```diff
@@ -0,0 +1,153 @@
...
+    }
+
+    /**
+     * Get the time since the last reputation.
+     *
+     * @param other receiver
+     * @return the time since the last vote in the requested time unit or  {@link Long#MAX_VALUE} if no entry was found.
+     */
+    public Duration getLastRatedDuration(User other) {
+        return getLastReputation(other).map(last -> Duration.between(last, Instant.now())).orElseGet(() -> Duration.ofDays(365));
```
[the remark this answers is not in the document]
Comment (RainbowDashLabs) [reply src/main/java/de/chojo/repbot/dao/access/guild/reputation/sub/RepUser.java:113]: True. Gonna adjust it to 1 year.

File: src/main/java/de/chojo/repbot/dao/access/guild/reputation/sub/RepUser.java [hunk 11]
```diff
@@ -0,0 +1,153 @@
...
+import org.slf4j.Logger;
+
+import javax.annotation.Nullable;
+import java.time.Duration;
+import java.time.Instant;
+import java.util.Optional;
+
+import static org.slf4j.LoggerFactory.getLogger;
+
+public class RepUser extends QueryFactoryHolder implements GuildHolder, MemberHolder {
```
[the remark this answers is not in the document]
Comment (RainbowDashLabs) [reply src/main/java/de/chojo/repbot/dao/access/guild/reputation/sub/RepUser.java:24]: True. Removed

File: src/main/java/de/chojo/repbot/dao/access/guild/settings/sub/Thanking.java (hunk 1 above)
Comment (RainbowDashLabs) [reply src/main/java/de/chojo/repbot/dao/access/guild/settings/sub/Thanking.java:37]: The constructor is used by the build function where the reaction actually can be null
Actually no. Probably just an artifact. Gonna pass the default value in there.
I also moved the default value to a constant.

Additionally I added the default value to the column to ensure it is always set.

File: src/main/java/de/chojo/repbot/listener/StateListener.java (hunk 4 above)
Comment (RainbowDashLabs) [reply src/main/java/de/chojo/repbot/listener/StateListener.java:58]: I fixed it. I will also create an issue to remove the migration implementation after a grace period for not migrated guilds

File: src/main/java/de/chojo/repbot/service/GdprService.java (hunk 5 above)
Comment (RainbowDashLabs) [reply src/main/java/de/chojo/repbot/service/GdprService.java:65]: I created another static method to create and directly execute removal tasks.

File: src/main/java/de/chojo/repbot/service/RepBotCachePolicy.java (hunk 6 above)
Comment (RainbowDashLabs) [reply src/main/java/de/chojo/repbot/service/RepBotCachePolicy.java:48]: Oh yeah...

File: src/main/resources/database/postgres/1/patch_17.sql (hunk 7 above)
Comment (RainbowDashLabs) [reply src/main/resources/database/postgres/1/patch_17.sql:5]: Applied reformatting

**Review by yannicklamprecht — APPROVED** [review]
On the first sight, looks good.

**Pull request merged**